pkg/controller/persistentvolume/persistentvolume_controller.go

```diff
@@ -17,20 +17,59 @@ limitations under the License.
 package persistentvolume
 
 import (
+	"fmt"
 	"time"
 
 	"k8s.io/kubernetes/pkg/api"
 	"k8s.io/kubernetes/pkg/client/cache"
 	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
 	"k8s.io/kubernetes/pkg/cloudprovider"
 	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/conversion"
 	"k8s.io/kubernetes/pkg/runtime"
 	"k8s.io/kubernetes/pkg/volume"
 	"k8s.io/kubernetes/pkg/watch"
 
 	"github.com/golang/glog"
 )
 
+// Design:
+//
+// The fundamental key to this design is the bi-directional "pointer" between
+// PersistentVolumes (PVs) and PersistentVolumeClaims (PVCs), which is
+// represented here as pvc.Spec.VolumeName and pv.Spec.ClaimRef. The bi-directionality
+// is complicated to manage in a transactionless system, but without it we
+// can't ensure sane behavior in the face of different forms of trouble.  For
+// example, a rogue HA controller instance could end up racing and making
+// multiple bindings that are indistinguishable, resulting in potential data
+// loss.
+//
+// This controller supports pre-bound (by the creator) objects in both
+// directions: a PVC that wants a specific PV or a PV that is reserved for a
+// specific PVC.
+//
+// The binding is two-step process. PV.Spec.ClaimRef is modified first and
+// PVC.Spec.VolumeName second. At any point of this transaction, the PV or PVC
+// can be modified by user or other controller or completelly deleted. Also, two
+// (or more) controllers may try to bind different volumes to different claims
+// at the same time. The controller must recover from any conflicts that may
+// arise from these conditions.
```

Comment:
> What does "bound by controller" mean?

Using this annotation on claims the controller remembers who set `claim.Spec.VolumeName`. On PVs, it says that the controller set `pv.Spec.ClaimRef.Name+NameSpace`.

If the annotation is present, we can set pv.Spec.ClaimRef to nil when a recycled volume is put back to `Available` phase. If it was the user who wanted specific claim for a PV, we must not touch it - recycled volume will be still "pre-bound" to a claim. It's quite different to current recycler, which makes pre-bound volumes available to anyone after recycling.

There are also other corner cases where this annotation is used to resolve conflicts, it's always deciding between "controller has chosen PV/PVC, therefore do action A" or "user explicitly requested specific PV/PVC and do B".

> What would be different in your example if pv2's `ReclaimPolicy=Retain`

As you said, pv2 will remain `Released` and does not play any role in binder decision. Nothing should bind to it ever, claim1 should be Pending forever.

> @Erin: Which takes precedence

There is no precedence, the binder will process claims and pvs as they are created and if two are created roughly at the same time then their order is not defined. And if both `pv.Spec.ClaimRef` and `claim.Spec.VolumeName` are set both must match. I think I explained what happens when `pv.Spec.ClaimRef` and `claim.Spec.VolumeName` won't match in [previous comment](https://github.com/kubernetes/kubernetes/pull/24331/files/5e1357376c07be134eda214d0d4851ca67a2d959..c572ba8afee569c7372c773e8bafcbac26e2fe9a#r61045350)